### PR TITLE
[BEAM-12074] Add @with_docs_from decorator for generating API docs

### PR DESCRIPTION
**Pull request by TheNeuralBit**
This PR adds the necessary infrastructure for generating API documentation for the Beam DataFrame API. Specifically, it adds `frame_base.with_docs_from`, which can be used to pull the docstring from a corresponding pandas method and modify it appropriately for the Beam DataFrame API. Specifically these modifications include:
- Any existing docstring on the method is written to a custom "Differences from pandas" section, included after the parameters and return type, and just before the "See also" section.
- The Examples section from pandas is included in the new docstring, but with a disclaimer that these are _pandas_ examples. We also add a header to make sure the sphinx doctest framework doesn't try to execute these examples. We already have an extensive framework for verifying our implementation against pandas' doctests.
- References to "version x.y" are updates to state "_pandas_ version x.y" to avoid confusion with beam version.
- Refernces to Series and DataFrame outside of the Examples section are replaced with DeferredSeries and DeferredDataFrame.

The PR also includes some modifications to `generate_pydoc.sh` to make it work with this new form of docstring:
- Allow the "Differences from pandas" section
- Make sure we include `inherited-members` for `apache_beam.dataframe.frames` (see [dev@ discussion](https://lists.apache.org/thread.html/r1718e082059f75a93e00634fe869fe0556212b82f23cbd0d8e73fd96%40%3Cdev.beam.apache.org%3E))
- Allow intersphinx to add links to numpy and pandas documentation
- Ignore WARNINGs related to `apache_beam.dataframe.frames`, unfortunately the pandas docstrings are not compatible with our "nitpicky" config and thus raise a lot of warnings: https://github.com/apache/beam/blob/476c2b1a39e4190b8391973772e4ab826a13f153/sdks/python/scripts/generate_pydoc.sh#L218

Remaining changes include:
- Update some docstrings for better rendering in the html API docs
- Add (skipped) DocstringTest to identify un-documented operations
- Demonstrate the use of `@with_docs_from` on a few operations.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>SDK</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by TheNeuralBit** [comment]
R: @pcoet 
Could you review the text of the module-level and method-level docstrings I've added here?

File: sdks/python/apache_beam/dataframe/frame_base.py [hunk 1]
```diff
@@ -356,6 +374,85 @@ def wrapper(*args, **kwargs):
...
+    'Raises',
+    BEAM_SPECIFIC,
+    'See Also',
+    'Notes',
+    'Examples'
+]
+
+EXAMPLES_DISCLAIMER = (
+    "**NOTE:** These examples are pulled directly from the pandas documentation "
+    "for convenience. The Beam DataFrame API will look different because it is "
```
Comment (pcoet) [inline sdks/python/apache_beam/dataframe/frame_base.py:391]: "The Beam DataFrame API" -> "Usage of the Beam DataFrame API"

File: sdks/python/apache_beam/dataframe/frames.py [hunk 2]
```diff
@@ -1215,17 +1267,22 @@ def aggregate(self, func, axis=0, *args, **kwargs):
...
 
   clip = frame_base._elementwise_method(
-      'clip', restrictions={'axis': lambda axis: axis in (0, 'index')})
+      'clip', restrictions={'axis': lambda axis: axis in (0, 'index')}, base=pd.DataFrame)
 
+  @frame_base.with_docs_from(pd.DataFrame)
   @frame_base.args_to_kwargs(pd.DataFrame)
   @frame_base.populate_defaults(pd.DataFrame)
   def corr(self, method, min_periods):
+    """Only ``method="pearson"`` can be parallelized, other methods require
```
Comment (pcoet) [inline sdks/python/apache_beam/dataframe/frames.py:1282]: "parallelized, other" -> "parallelized. Other"

File: sdks/python/apache_beam/dataframe/__init__.py [hunk 3]
```diff
@@ -14,4 +14,17 @@
...
+
+- For high-level documentation see
+  https://beam.apache.org/documentation/dsls/dataframes/overview/
+- :mod:`apache_beam.dataframe.io`: DataFrame I/Os
+- :mod:`apache_beam.dataframe.frames`: DataFrame operations
+- :mod:`apache_beam.dataframe.convert`: Conversion between
+  :class:`~apache_beam.pvalue.PCollection` and
+  :class:`~apache_beam.dataframe.frames.DeferredDataFrame`.
+- :mod:`apache_beam.dataframe.transforms`: Embed DataFrame operations in a
+  Beam pipeline.
```
Comment (pcoet) [inline sdks/python/apache_beam/dataframe/__init__.py:27]: "Embed DataFrame operations in a Beam pipeline." -> "DataFrame operations for use in a Beam pipeline."

**Review by pcoet — APPROVED** [review]
LGTM. I made a few comments, but they're super nitty.

File: sdks/python/apache_beam/dataframe/frame_base.py (hunk 1 above)
Comment (TheNeuralBit) [reply sdks/python/apache_beam/dataframe/frame_base.py]: Done

File: sdks/python/apache_beam/dataframe/frames.py (hunk 2 above)
Comment (TheNeuralBit) [reply sdks/python/apache_beam/dataframe/frames.py]: Done

File: sdks/python/apache_beam/dataframe/__init__.py (hunk 3 above)
Comment (TheNeuralBit) [reply sdks/python/apache_beam/dataframe/__init__.py]: Done. Thank you!

**Comment by TheNeuralBit** [comment]
R: @rohdesamuel would you be able to review this?
If it helps I could stage a copy of the API docs generated from this PR

**Review by rohdesamuel — APPROVED** [review]

**Comment by TheNeuralBit** [comment]
GHA failures look like flakes (BEAM-12178, BEAM-12163)

**Pull request merged**